zerver/lib/test_runner.py

```diff
@@ -472,9 +476,49 @@ def run_tests(self, test_labels: List[str],
         result = self.run_suite(suite)
         self.teardown_test_environment()
         failed = self.suite_result(suite, result)
+        failed_thread = []  # type: List[str]
         if not failed:
             write_instrumentation_reports(full_suite=full_suite, include_webhooks=include_webhooks)
-        return failed, result.failed_tests
+        else:
+            # Below is the logic for passing the list of tests ran in a failed
+            # worker, for use in the --replicate option when running test-backend.
+            #
+            # If we encounter a failure we verify that the ParallelTestSuite
+            # is being used too ensure subsuites exist.
+            if self.parallel == 1:
+                return failed, result.failed_tests, failed_thread
+
+            global _subsuite_index_to_pid
+            args = [
+                (index, subsuite[1])
+                for index, subsuite in enumerate(suite.subsuites)
+            ]  # type: List[Tuple[int, List[str]]]
+
+            # We iterate through the subsuites for the failure.
+            # When we find it, assign the index for use in accumulating the
+            # individual tests.
+            index = -1
+            for indexed_tests in args:
+                for test in indexed_tests[1]:
+                    if result.failed_tests[0] == test:
+                        index = indexed_tests[0]
+                        break
+            assert index is not -1
+
+            # We now check which subsuites ran in the same thread as the
+            # subsuite containing the failed test.
+            failed_indices = []  # type: List[int]
+            for i in range(len(_subsuite_index_to_pid)):
+                if _subsuite_index_to_pid[i] == _subsuite_index_to_pid[index]:
+                    failed_indices.append(i)
+
+            # Finally, add all the tests from the subsuites that were ran in
+            # the failed thread.
+            for subsuite in args:
+                for index in failed_indices:
+                    if subsuite[0] == index:
+                        failed_thread.extend(subsuite[1])
```

Comment:
Simpler, and not quadratic-time:

```python
failed_pid = _subsuite_index_to_pid[suite.subsuites.index(result.failed_tests[0])]
failed_thread = [
    test
    for subsuite, pid in zip(suite.subsuites, _subsuite_index_to_pid)
    if pid == failed_pid
    for test in subsuite
]
```

(Although why don’t we just output the list of subsuites instead of the list of tests?)

Reply:
Ah very cool, I'll study this (I'm trying to get in the habit of thinking more pythonically, and this is great).  As for a using a list of tests, can we pass a list of subsuites to `test-backend` and still have it run the set of tests with the existing logic? I was sort of piggy backing on the way `result.failed_tests` is used.

Reply:
Yeah you can run, e.g., `tools/test-backend zerver.tests.test_alert_words.AlertWordTests zerver.tests.test_archive.GlobalPublicStreamTest`. Although since the subsuite name `zerver.tests.test_alert_words.AlertWordTests` doesn’t seem to be stored in the subsuite, getting the subsuite names may be too hard.

Reply:
For the line
```
failed_pid = _subsuite_index_to_pid[suite.subsuites.index(result.failed_tests[0])]
```
I'm also seeing that the name of the failed test wasn't in the list.